src/cache/cache_las.c

```diff
@@ -1101,65 +1110,85 @@ __wt_las_sweep(WT_SESSION_IMPL *session)
 
-            /*
-             * Expect an update entry with:
-             *	1. not in a prepare locked state
-             *	2. durable timestamp as not max timestamp.
-             *	3. for an in-progress prepared update, durable
-             *	timestamp should be zero.
-             *	4. no restriction on durable timestamp value
-             *	for other updates.
-             */
-            WT_ASSERT(session, prepare_state != WT_PREPARE_LOCKED &&
-                durable_timestamp != WT_TS_MAX &&
-                (prepare_state != WT_PREPARE_INPROGRESS || durable_timestamp == 0));
+        /*
+         * Check to see if the page or key has changed this iteration, and if they have, setup
+         * context for safely removing obsolete updates.
+         */
+        remove_entry = false;
+        if (prev_entry_can_remove) {
+            if (las_btree_id != saved_btree_id || saved_key->size != las_key.size ||
+              memcmp(saved_key->data, las_key.data, las_key.size) != 0) {
+                remove_entry = true;
+                saved_btree_id = las_btree_id;
+                WT_ERR(__wt_buf_set(session, saved_key, las_key.data, las_key.size));
+            } else {
+                if (upd_type != WT_UPDATE_MODIFY)
+                    remove_entry = true;
+                else
+                    skip_update_modify_cnt++;
+            }
+        }
 
-            WT_ASSERT(session,
-              (prepare_state == WT_PREPARE_INPROGRESS || durable_timestamp >= las_timestamp));
+        /* Remove the obsolete entries */
+        if (remove_entry) {
+            /* Rewind the cursor to a point of remove record. */
+            for (i = 0; i < skip_update_modify_cnt + 1; i++)
+                cursor->prev(cursor);
+
+            WT_ERR(cursor->get_key(cursor, &las_btree_id, &las_key, &las_timestamp, &las_txnid));
+            WT_ERR(
+              cursor->get_value(cursor, &durable_timestamp, &prepare_state, &upd_type, &las_value));
+
+            /* Remove each individual LAS record till the current record. */
+            for (i = 0; i < skip_update_modify_cnt + 1; i++) {
+                __wt_verbose(session, WT_VERB_LOOKASIDE_ACTIVITY,
+                  "Sweep removing lookaside entry with "
+                  "btree ID: %" PRIu32 " saved key size: %" WT_SIZET_FMT ", record type: %" PRIu8
+                  " transaction ID: %" PRIu64,
+                  las_btree_id, saved_key->size, upd_type, las_txnid);
+
+                cursor->remove(cursor);
+                cursor->next(cursor);
+                WT_ERR(
+                  cursor->get_key(cursor, &las_btree_id, &las_key, &las_timestamp, &las_txnid));
+                WT_ERR(cursor->get_value(
+                  cursor, &durable_timestamp, &prepare_state, &upd_type, &las_value));
+                ++remove_cnt;
+            }
+            skip_update_modify_cnt = 0;
+        }
```

Comment:
I'm not sure what the rest of sweep looks like but at least for the changes here, scanning with `prev` should be easier. There just needs to be a flag to track whether the last record we saw was a modify. And if so, a standard update should be protected from removal even if it is expired.

There's probably some nuance that I'm missing but I still imagine it's simpler than this. Not sure whether it's worth the effort to try this but that's my initial thought.

Reply:
I fix the removal when it is travesing backwards.

@tetsuo-cpp, I understand your comment as traverse from backwards instead of forward, is it correct?
I thought of writing the code using the `prev` instead of `next`, but what I felt is that using of `next` there is a greater chance of removing the records than using the `prev` when there is a hight amount writes that are happening to the LAS.